File: Level.1/소수찾기.cpp
#include <string>
#include <vector>
#include <map>
using namespace std;

bool visited[7]={0,};
char num[7];
int M;
string temp = "";
map<int, int> mp;

bool is_prime(int p) {
    if(p < 2) return false;
    for(int i = 2; i * i <= p; ++i) {
        if(p % i == 0) return false;
    }
    return true;
}

void dfs(int height) {
    if(height == M) return;
    
    for(int i = 0; i < M; ++i) {
        if(!visited[i]) {
            visited[i] = true;
            temp += num[i];
            ++mp[stoi(temp)];
            dfs(height+1);
            temp = temp.substr(0, height);
            visited[i] = false;
        }
    }
}

int solution(string numbers) {
    int answer = 0;
    M = numbers.length();
    for(int i = 0; i < M; ++i) {
        num[i] = numbers[i];
    }
    dfs(0);
    map<int, int>::iterator iter;
    for(iter = mp.begin(); iter != mp.end(); ++iter) {
        if(is_prime(iter->first)) ++answer;
    }
    return answer;
}

File: JavaScript/Level.2/소수찾기.js
function isPrime(p) {
    if (p < 2) return false;
    let sqrt = Math.floor(Math.sqrt(p));
    for (let i = 2; i <= sqrt; ++i) {
        if (p % i === 0) {
            return false;
        }
    }
    return true;
}

function Permutation(numbers) {
    let n = numbers.length;
    let visited = Array.from({ length: n });
    let choose = "";
    let answer = [];
    function PermutationUtil(depth, r) {
        if (depth === r) {
            let num = Number(choose);
            if (!answer.includes(num)) {
                answer.push(num);
            }
            return;
        }
        for (let i = 0; i < n; ++i) {
            if (!visited[i]) {
                visited[i] = 1;
                choose += String(numbers[i]);
                PermutationUtil(depth + 1, r);
                visited[i] = 0;
                choose = choose.substr(0, choose.length - 1);
            }
        }
    }
    for (let r = 0; r <= n; ++r) {
        PermutationUtil(0, r);
    }
    return answer;
}

function solution(numbers) {
    let answer = 0;
    let P = Permutation(numbers);
    for (let p of P) {
        if (isPrime(p)) {
            ++answer;
        }
    }
    return answer;
}
